nearcache_map.go

```diff
@@ -0,0 +1,265 @@
+)
+
+type nearCacheMap struct {
+	nc             *nearCache
+	ncc            *nearcache.Config
+	toNearCacheKey func(key interface{}) (interface{}, error)
+	ss             *serialization.Service
+}
+
+func newNearCacheMap(nc *nearCache, ncc *nearcache.Config, ss *serialization.Service) (nearCacheMap, error) {
+	ncm := nearCacheMap{
+		nc:  nc,
+		ncc: ncc,
+		ss:  ss,
+	}
+	// the only valid local policy on the client side is invalidate.
+	ncm.registerInvalidationListener()
+	if ncc.Preloader.Enabled {
+		if err := ncm.preload(); err != nil {
+			return nearCacheMap{}, fmt.Errorf("preloading near cache: %w", err)
+		}
+	}
+	// toNearCacheKey returns the raw key if SerializeKeys is not true.
+	if ncc.SerializeKeys {
+		ncm.toNearCacheKey = func(key interface{}) (interface{}, error) {
+			data, err := ss.ToData(key)
+			if err != nil {
+				return nil, err
+			}
+			// byte slices cannot be map keys
+			return data, nil
+		}
+	} else {
+		ncm.toNearCacheKey = func(key interface{}) (interface{}, error) {
+			return key, nil
+		}
+	}
+	return ncm, nil
+}
+
+func (ncm *nearCacheMap) registerInvalidationListener() {
+	fmt.Println("IMPLEMENT ME: registerInvalidationListener")
+}
+
+func (ncm *nearCacheMap) preload() error {
+	panic("implement me!")
+}
+
+func (ncm *nearCacheMap) ContainsKey(ctx context.Context, key interface{}, m *Map) (found bool, err error) {
+	key, err = ncm.toNearCacheKey(key)
+	if err != nil {
+		return false, err
+	}
+	cached, ok, err := ncm.getCachedValue(key, false)
+	if err != nil {
+		return false, err
+	}
+	if ok {
+		return cached != nil, nil
+	}
+	keyData, err := m.validateAndSerialize(key)
+	if err != nil {
+		return false, err
+	}
+	return m.containsKeyFromRemote(ctx, keyData)
+}
+
+func (ncm *nearCacheMap) Delete(ctx context.Context, m *Map, key interface{}) error {
+	key, err := ncm.toNearCacheKey(key)
+	if err != nil {
+		return err
+	}
+	defer ncm.nc.Invalidate(key)
+	keyData, err := m.validateAndSerialize(key)
+	if err != nil {
+		return err
+	}
+	return m.deleteFromRemote(ctx, keyData)
+}
+
+func (ncm *nearCacheMap) Get(ctx context.Context, m *Map, key interface{}) (interface{}, error) {
+	key, err := ncm.toNearCacheKey(key)
+	if err != nil {
+		return nil, err
+	}
+	cached, found, err := ncm.getCachedValue(key, false)
+	if err != nil {
+		return nil, err
+	}
+	if found {
+		return cached, nil
+	}
+	// value not found in local cache.
+	// get it from remote.
+	value, err := ncm.getFromRemote(ctx, m, key)
+	if err != nil {
+		ncm.nc.Invalidate(key)
+		return nil, err
+	}
+	return value, nil
+}
+
+func (ncm *nearCacheMap) Put(ctx context.Context, m *Map, key, value interface{}, ttl int64) (interface{}, error) {
+	key, err := ncm.toNearCacheKey(key)
+	if err != nil {
+		return nil, err
+	}
+	defer ncm.nc.Invalidate(key)
+	keyData, valueData, err := m.validateAndSerialize2(key, value)
+	if err != nil {
+		return false, err
+	}
+	return m.putWithTTLFromRemote(ctx, keyData, valueData, ttl)
```

Comment:
Why do we return false if err is not nil? It should be nil I guess. Also why not use `m.putWithTTL`?

Reply:
`m.Put` and `m.PutWithTTL` already calls this method if the near cache is enabled. If there is an error, the other returned value is not used, so it's not very important what value is there. But conventionally, the zero-value is returned.

Reply:
Refactored at: https://github.com/hazelcast/hazelcast-go-client/pull/806/commits/a5b6adc0848f374f5049cf9a75bcb2aacccf9699